### PR TITLE
ddl: support add foreign key constraint with auto-create index

### LINKED ISSUE #38762: support add foreign key constraint with auto-create index.
## Dev Task

support add foreign key constraint with auto-create index.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

Issue Number: close #38762

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -6276,6 +6276,24 @@ func (d *ddl) CreateForeignKey(ctx sessionctx.Context, ti ast.Ident, fkName mode
...
+	if model.FindIndexByColumns(t.Meta(), fkInfo.Cols...) == nil {
+		// Need to auto create index for fk cols
+		if ctx.GetSessionVars().StmtCtx.MultiSchemaInfo == nil {
+			ctx.GetSessionVars().StmtCtx.MultiSchemaInfo = model.NewMultiSchemaInfo()
+		}
+		indexPartSpecifications := make([]*ast.IndexPartSpecification, 0, len(fkInfo.Cols))
+		for _, col := range fkInfo.Cols {
+			indexPartSpecifications = append(indexPartSpecifications, &ast.IndexPartSpecification{
+				Column: &ast.ColumnName{Name: col},
+				Length: -1,
```
Comment (tangenta) [inline ddl/ddl_api.go:6288]:
```suggestion
				Length: types.UnspecifiedLength, // Index prefixes on foreign key columns are not supported.
```

File: ddl/foreign_key_test.go [hunk 2]
```diff
@@ -1426,6 +1424,26 @@ func TestAddForeignKey(t *testing.T) {
 	tk.MustExec("alter table t1 add index(b)")
 	tk.MustExec("alter table t2 add foreign key (b) references t1(b);")
 	tk.MustGetDBError("alter table t2 add foreign key (b) references t2(b);", infoschema.ErrCannotAddForeign)
+	// Test auto-create index when create foreign key constraint.
```
Comment (tangenta) [inline ddl/foreign_key_test.go:1427]: Could you add a test for creating multiple foreign keys?

Comment (crazycs520) [reply ddl/foreign_key_test.go:1428]: nice catch , done.

**Review by tangenta — APPROVED** [review]

File: ddl/foreign_key.go [hunk 3]
```diff
@@ -666,6 +655,16 @@ func checkAddForeignKeyValidInOwner(d *ddlCtx, t *meta.Meta, schema string, tbIn
...
 	}
+	// check foreign key columns should have index.
+	if len(fk.Cols) == 1 && tbInfo.PKIsHandle {
+		pkCol := tbInfo.GetPkColInfo()
+		if pkCol != nil && pkCol.Name.L == fk.Cols[0].L {
+			return nil
+		}
+	}
+	if model.FindIndexByColumns(tbInfo, fk.Cols...) == nil {
+		return errors.Errorf("Failed to add the foreign key constraint. Missing index for '%s' foreign key columns in the table '%s'", fk.Name, tbInfo.Name)
```
Comment (wjhuang2016) [inline ddl/foreign_key.go:666]: Is there any error code for it in MySQL.

Comment (crazycs520) [reply ddl/foreign_key.go:666]: No, MySQL doesn't have this error. This is a slight compatibility issue with MySQL.
But the error should only happen when concurrently executing `drop index` and  `add foreign key`.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**